google-cloud-pubsublite/src/main/java/com/google/cloud/pubsublite/internal/TopicBacklogReaderImpl.java

```diff
@@ -0,0 +1,102 @@
+import com.google.protobuf.Timestamp;
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public final class TopicBacklogReaderImpl implements TopicBacklogReader {
+  private final TopicStatsClient client;
+  private final TopicPath topicPath;
+
+  public TopicBacklogReaderImpl(TopicStatsClient client, TopicPath topicPath) {
+    this.client = client;
+    this.topicPath = topicPath;
+  }
+
+  private static Optional<Instant> minTimestamp(Optional<Instant> t1, Instant t2) {
+    if (!t1.isPresent() || t2.isBefore(t1.get())) {
+      return Optional.of(t2);
+    }
+    return t1;
+  }
+
+  private static Instant toJavaInstant(Timestamp timestamp) {
+    return Instant.ofEpochSecond(timestamp.getSeconds(), timestamp.getNanos());
+  }
+
+  private static Timestamp toProtoTimestamp(Instant instant) {
+    return Timestamp.newBuilder()
+        .setSeconds(instant.getEpochSecond())
+        .setNanos(instant.getNano())
+        .build();
+  }
+
+  @Override
+  public ApiFuture<ComputeMessageStatsResponse> computeMessageStats(
+      Map<Partition, Offset> subscriptionState) {
+    List<ApiFuture<ComputeMessageStatsResponse>> perPartitionFutures =
+        subscriptionState.entrySet().stream()
+            .map(
+                e ->
+                    client.computeMessageStats(
+                        topicPath, e.getKey(), e.getValue(), Offset.of(Integer.MAX_VALUE)))
+            .collect(Collectors.toList());
+    return ApiFutures.transformAsync(
+        ApiFutures.allAsList(perPartitionFutures),
+        responses -> {
+          Optional<Instant> minPublishTime = Optional.empty();
+          Optional<Instant> minEventTime = Optional.empty();
+          long messageBytes = 0;
+          long messageCount = 0;
+          for (ComputeMessageStatsResponse response : responses) {
+            messageBytes += response.getMessageBytes();
+            messageCount += response.getMessageCount();
+            if (response.hasMinimumPublishTime()) {
+              minPublishTime =
+                  minTimestamp(minPublishTime, toJavaInstant(response.getMinimumPublishTime()));
+            }
+            if (response.hasMinimumEventTime()) {
+              minEventTime =
+                  minTimestamp(minPublishTime, toJavaInstant(response.getMinimumEventTime()));
+            }
+          }
+          ComputeMessageStatsResponse.Builder builder =
+              ComputeMessageStatsResponse.newBuilder()
+                  .setMessageBytes(messageBytes)
+                  .setMessageCount(messageCount);
+          if (minPublishTime.isPresent()) {
+            builder.setMinimumPublishTime(toProtoTimestamp(minPublishTime.get()));
+          }
+          if (minEventTime.isPresent()) {
+            builder.setMinimumEventTime(toProtoTimestamp(minEventTime.get()));
+          }
+          return ApiFutures.immediateFuture(builder.build());
```

Comment:
this should just be .transform above if the future you return is immediately done.

Reply:
Oh, yeah, I had some issues with transform before I realized you could use a lambda instead of implementing the ApiFunction class, but this great